File: tests/wasm/test_ehlers_itrend.js
/**
 * WASM binding tests for Ehlers ITrend indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('Ehlers ITrend partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    
    
    const result = wasm.ehlers_itrend_js(close, 12, 50);
    assert.strictEqual(result.length, close.length);
    
    
    const resultCustomWarmup = wasm.ehlers_itrend_js(close, 15, 50);
    assert.strictEqual(resultCustomWarmup.length, close.length);
    
    const resultCustomMaxDc = wasm.ehlers_itrend_js(close, 12, 40);
    assert.strictEqual(resultCustomMaxDc.length, close.length);
});

test('Ehlers ITrend accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.ehlersItrend;
    
    const result = wasm.ehlers_itrend_js(
        close,
        expected.defaultParams.warmupBars,
        expected.defaultParams.maxDcPeriod
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        0.1,  
        "Ehlers ITrend last 5 values mismatch"
    );
    
    
    await compareWithRust('ehlers_itrend', result, 'close', expected.defaultParams);
});

test('Ehlers ITrend empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.ehlers_itrend_js(empty, undefined, undefined);
    }, /Input data is empty/);
});

test('Ehlers ITrend all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.ehlers_itrend_js(allNaN, undefined, undefined);
    }, /All values are NaN/);
});

test('Ehlers ITrend insufficient data', () => {
    
    const smallData = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);
    
    assert.throws(() => {
        wasm.ehlers_itrend_js(smallData, 10, 50);
    }, /Not enough data for warmup/);
});

test('Ehlers ITrend zero warmup', () => {
    
    const close = new Float64Array(testData.close);
    
    assert.throws(() => {
        wasm.ehlers_itrend_js(close, 0, 50);
    }, /Invalid warmup_bars/);
});

test('Ehlers ITrend invalid max_dc', () => {
    
    const close = new Float64Array(testData.close);
    
    assert.throws(() => {
        wasm.ehlers_itrend_js(close, 12, 0);
    }, /Invalid max_dc_period/);
});

test('Ehlers ITrend reinput', () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.ehlersItrend;
    const params = expected.defaultParams;
    
    
    const firstResult = wasm.ehlers_itrend_js(
        close,
        params.warmupBars,
        params.maxDcPeriod
    );
    assert.strictEqual(firstResult.length, close.length);
    
    
    const secondResult = wasm.ehlers_itrend_js(
        firstResult,
        params.warmupBars,
        params.maxDcPeriod
    );
    assert.strictEqual(secondResult.length, firstResult.length);
    
    
    
    
    
    if (secondResult.length > 30) {
        let hasValidValues = false;
        for (let i = 30; i < secondResult.length; i++) {
            if (!isNaN(secondResult[i])) {
                hasValidValues = true;
                break;
            }
        }
        assert(hasValidValues, 'Expected some valid values after warmup and lookback period');
    }
});

test('Ehlers ITrend NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.ehlersItrend;
    const params = expected.defaultParams;
    const warmupBars = params.warmupBars;
    
    const result = wasm.ehlers_itrend_js(
        close,
        params.warmupBars,
        params.maxDcPeriod
    );
    assert.strictEqual(result.length, close.length);
    
    
    for (let i = 0; i < warmupBars; i++) {
        assert(isNaN(result[i]), `Expected NaN at warmup index ${i}`);
    }
    
    
    if (result.length > warmupBars) {
        for (let i = warmupBars; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
    
    
    assert.strictEqual(warmupBars, 12, `Expected warmupBars=12, got ${warmupBars}`);
});

test('Ehlers ITrend batch single parameter set', () => {
    
    
    if (typeof wasm.ehlers_itrend_batch === 'undefined') {
        console.log('  Skipping: Batch functions not implemented for ehlers_itrend');
        return;
    }
    
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    
    const batchResult = wasm.ehlers_itrend_batch(close, {
        warmup_range: [12, 12, 0],
        max_dc_range: [50, 50, 0]
    });
    
    
    const singleResult = wasm.ehlers_itrend_js(close, 12, 50);
    
    assert.strictEqual(batchResult.values.length, singleResult.length);
    assertArrayClose(batchResult.values, singleResult, 1e-10, "Batch vs single mismatch");
});

test('Ehlers ITrend batch multiple parameters', () => {
    
    if (typeof wasm.ehlers_itrend_batch === 'undefined') {
        console.log('  Skipping: Batch functions not implemented');
        return;
    }
    
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    
    const batchResult = wasm.ehlers_itrend_batch(close, {
        warmup_range: [10, 14, 2],    
        max_dc_range: [40, 50, 10]    
    });
    
    
    assert.strictEqual(batchResult.combos.length, 6);
    assert.strictEqual(batchResult.values.length, 6 * 100);
    
    
    const firstRow = batchResult.values.slice(0, 100);
    const singleResult = wasm.ehlers_itrend_js(close, 10, 40);
    assertArrayClose(
        firstRow, 
        singleResult, 
        1e-10, 
        "First batch row mismatch"
    );
});

test('Ehlers ITrend batch metadata', () => {
    
    if (typeof wasm.ehlers_itrend_batch === 'undefined') {
        console.log('  Skipping: Batch metadata functions not implemented');
        return;
    }
    
    
    const close = new Float64Array(20).fill(100);
    const result = wasm.ehlers_itrend_batch(close, {
        warmup_range: [10, 14, 2],    
        max_dc_range: [40, 50, 10]    
    });
    
    
    assert.strictEqual(result.combos.length, 6);
    
    
    assert.strictEqual(result.combos[0].warmup_bars, 10);
    assert.strictEqual(result.combos[0].max_dc_period, 40);
    
    
    assert.strictEqual(result.combos[1].warmup_bars, 10);
    assert.strictEqual(result.combos[1].max_dc_period, 50);
    
    
    assert.strictEqual(result.combos[2].warmup_bars, 12);
    assert.strictEqual(result.combos[2].max_dc_period, 40);
});

test('Ehlers ITrend batch warmup validation', () => {
    
    if (typeof wasm.ehlers_itrend_batch === 'undefined') {
        console.log('  Skipping: Batch functions not implemented');
        return;
    }
    
    
    const close = new Float64Array(testData.close.slice(0, 30));
    
    const batchResult = wasm.ehlers_itrend_batch(close, {
        warmup_range: [10, 15, 5],    
        max_dc_range: [50, 50, 0]     
    });
    
    assert.strictEqual(batchResult.combos.length, 2);
    
    
    for (let combo = 0; combo < batchResult.combos.length; combo++) {
        const warmupBars = batchResult.combos[combo].warmup_bars;
        const rowStart = combo * 30;
        const rowData = batchResult.values.slice(rowStart, rowStart + 30);
        
        
        for (let i = 0; i < warmupBars; i++) {
            assert(isNaN(rowData[i]), `Expected NaN at index ${i} for warmup_bars=${warmupBars}`);
        }
    }
});

test('Ehlers ITrend batch edge cases', () => {
    
    if (typeof wasm.ehlers_itrend_batch === 'undefined') {
        console.log('  Skipping: Batch functions not implemented');
        return;
    }
    
    
    const close = new Float64Array(20).fill(0).map((_, i) => i + 1);
    
    
    const singleBatch = wasm.ehlers_itrend_batch(close, {
        warmup_range: [10, 10, 1],
        max_dc_range: [30, 30, 1]
    });
    
    assert.strictEqual(singleBatch.combos.length, 1);
    assert.strictEqual(singleBatch.values.length, 20);
    
    
    const zeroStepBatch = wasm.ehlers_itrend_batch(close, {
        warmup_range: [12, 12, 0],
        max_dc_range: [50, 50, 0]
    });
    
    assert.strictEqual(zeroStepBatch.combos.length, 1);
    assert.strictEqual(zeroStepBatch.values.length, 20);
    
    
    assert.throws(() => {
        wasm.ehlers_itrend_batch(new Float64Array([]), {
            warmup_range: [12, 12, 0],
            max_dc_range: [50, 50, 0]
        });
    }, /Input data is empty|All values are NaN/);
});

test('Ehlers ITrend batch performance test', () => {
    
    if (typeof wasm.ehlers_itrend_batch === 'undefined') {
        console.log('  Skipping: Batch functions not implemented');
        return;
    }
    
    
    const close = new Float64Array(testData.close.slice(0, 200));
    
    
    const startBatch = Date.now();
    const batchResult = wasm.ehlers_itrend_batch(close, {
        warmup_range: [10, 20, 2],    
        max_dc_range: [40, 50, 5]     
    });
    const batchTime = Date.now() - startBatch;
    
    
    const startSingle = Date.now();
    const singleResults = [];
    for (let warmup = 10; warmup <= 20; warmup += 2) {
        for (let maxDc = 40; maxDc <= 50; maxDc += 5) {
            singleResults.push(...wasm.ehlers_itrend_js(close, warmup, maxDc));
        }
    }
    const singleTime = Date.now() - startSingle;
    
    
    assert.strictEqual(batchResult.values.length, singleResults.length);
    
    
    console.log(`  Ehlers ITrend Batch time: ${batchTime}ms, Single calls time: ${singleTime}ms`);
});


test('Ehlers ITrend zero-copy API', () => {
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
    const warmupBars = 12;
    const maxDcPeriod = 50;
    
    
    if (!wasm.ehlers_itrend_alloc || !wasm.ehlers_itrend_into || !wasm.ehlers_itrend_free) {
        console.log('  Skipping: Zero-copy API not implemented for Ehlers ITrend');
        return;
    }
    
    
    const ptr = wasm.ehlers_itrend_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');
    
    
    const memoryBuffer = wasm.__wasm.memory.buffer;
    const memView = new Float64Array(
        memoryBuffer,
        ptr,
        data.length
    );
    
    
    memView.set(data);
    
    
    try {
        wasm.ehlers_itrend_into(ptr, ptr, data.length, warmupBars, maxDcPeriod);
        
        
        const regularResult = wasm.ehlers_itrend_js(data, warmupBars, maxDcPeriod);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue; 
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {
        
        wasm.ehlers_itrend_free(ptr, data.length);
    }
});

test('Ehlers ITrend batch with ergonomic API', () => {
    
    if (typeof wasm.ehlers_itrend_batch === 'undefined') {
        console.log('  Skipping: Ergonomic batch API not implemented for Ehlers ITrend');
        return;
    }
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    try {
        
        const result = wasm.ehlers_itrend_batch(close, {
            warmup_range: [10, 14, 2],     
            max_dc_range: [40, 50, 10]     
        });
        
        
        assert(result.values, 'Should have values array');
        assert(result.combos, 'Should have combos array');
        assert(typeof result.rows === 'number', 'Should have rows count');
        assert(typeof result.cols === 'number', 'Should have cols count');
        
        
        assert.strictEqual(result.rows, 6);
        assert.strictEqual(result.cols, 50);
        assert.strictEqual(result.combos.length, 6);
        assert.strictEqual(result.values.length, 300);
        
        
        for (let i = 0; i < result.combos.length; i++) {
            const combo = result.combos[i];
            const rowStart = i * 50;
            const rowData = result.values.slice(rowStart, rowStart + 50);
            
            const singleResult = wasm.ehlers_itrend_js(close, combo.warmupBars || combo.warmup_bars, combo.maxDcPeriod || combo.max_dc_period);
            assertArrayClose(
                rowData,
                singleResult,
                1e-10,
                `Combo ${i} mismatch (warmup=${combo.warmupBars || combo.warmup_bars}, maxDc=${combo.maxDcPeriod || combo.max_dc_period})`
            );
        }
    } catch (error) {
        
        
        console.log(`  Skipping: Ergonomic batch API error: ${error}`);
        return;
    }
});

test('Ehlers ITrend SIMD128 consistency', () => {
    
    
    const testCases = [
        { size: 20, warmup: 12 },
        { size: 100, warmup: 20 },
        { size: 1000, warmup: 50 }
    ];
    
    for (const testCase of testCases) {
        const data = new Float64Array(testCase.size);
        for (let i = 0; i < testCase.size; i++) {
            data[i] = Math.sin(i * 0.1) + Math.cos(i * 0.05);
        }
        
        const result = wasm.ehlers_itrend_js(data, testCase.warmup, 50);
        
        
        assert.strictEqual(result.length, data.length);
        
        
        for (let i = 0; i < testCase.warmup && i < data.length; i++) {
            assert(isNaN(result[i]), 
                   `Expected NaN at warmup index ${i} for size=${testCase.size}`);
        }
        
        
        let sumAfterWarmup = 0;
        let countAfterWarmup = 0;
        for (let i = testCase.warmup; i < result.length; i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} for size=${testCase.size}`);
            sumAfterWarmup += result[i];
            countAfterWarmup++;
        }
        
        
        if (countAfterWarmup > 0) {
            const avgAfterWarmup = sumAfterWarmup / countAfterWarmup;
            assert(Math.abs(avgAfterWarmup) < 100, `Average value ${avgAfterWarmup} seems unreasonable`);
        }
    }
});

test('Ehlers ITrend memory management', () => {
    
    if (!wasm.ehlers_itrend_alloc || !wasm.ehlers_itrend_free) {
        console.log('  Skipping: Zero-copy API not implemented');
        return;
    }
    
    
    const sizes = [100, 1000, 10000];
    
    for (const size of sizes) {
        const ptr = wasm.ehlers_itrend_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memoryBuffer = wasm.__wasm.memory.buffer;
        const memView = new Float64Array(memoryBuffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.ehlers_itrend_free(ptr, size);
    }
});

test('Ehlers ITrend batch comprehensive parameter sweep', () => {
    
    if (typeof wasm.ehlers_itrend_batch === 'undefined') {
        console.log('  Skipping: Batch functions not implemented');
        return;
    }
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    
    const batchResult = wasm.ehlers_itrend_batch(close, {
        warmup_range: [10, 20, 5],     
        max_dc_range: [30, 60, 15]     
    });
    
    
    assert.strictEqual(batchResult.combos.length, 9);
    assert.strictEqual(batchResult.values.length, 9 * 100);
    
    
    const expectedCombos = [];
    for (let warmup = 10; warmup <= 20; warmup += 5) {
        for (let maxDc = 30; maxDc <= 60; maxDc += 15) {
            expectedCombos.push({ warmup, maxDc });
        }
    }
    
    for (let i = 0; i < expectedCombos.length; i++) {
        const expected = expectedCombos[i];
        const actual = batchResult.combos[i];
        
        assert.strictEqual(actual.warmup_bars, expected.warmup, 
                          `Combo ${i} warmup mismatch`);
        assert.strictEqual(actual.max_dc_period, expected.maxDc, 
                          `Combo ${i} maxDc mismatch`);
        
        
        const rowStart = i * 100;
        const rowData = batchResult.values.slice(rowStart, rowStart + 100);
        const singleResult = wasm.ehlers_itrend_js(close, expected.warmup, expected.maxDc);
        
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Combo ${i} (warmup=${expected.warmup}, maxDc=${expected.maxDc}) calculation mismatch`
        );
    }
});



test.after(() => {
    console.log('Ehlers ITrend WASM tests completed');
});

File: tests/python/test_ehlers_itrend.py
"""
Python binding tests for Ehlers ITrend indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import numpy as np
import pytest
from test_utils import (
    load_test_data,
    assert_close,
    EXPECTED_OUTPUTS
)
from rust_comparison import compare_with_rust


import my_project


def assert_no_nan(arr, msg=""):
    """Assert no NaN values in array"""
    if np.any(np.isnan(arr)):
        raise AssertionError(f"{msg}: Found NaN values in array")

def assert_all_nan(arr, msg=""):
    """Assert all values are NaN"""
    if not np.all(np.isnan(arr)):
        raise AssertionError(f"{msg}: Not all values are NaN")


class TestEhlersITrend:
    def setup_method(self):
        """Load test data before each test."""
        self.data = load_test_data()
        self.close = np.array(self.data['close'], dtype=np.float64)
    
    def test_ehlers_itrend_partial_params(self):
        """Test with default parameters - mirrors check_itrend_partial_params"""
        
        
        result = my_project.ehlers_itrend(self.close, 12, 50)
        assert len(result) == len(self.close)
        
        
        result = my_project.ehlers_itrend(self.close, 15, 50)
        assert len(result) == len(self.close)
        
        result = my_project.ehlers_itrend(self.close, 12, 40)
        assert len(result) == len(self.close)
    
    def test_ehlers_itrend_accuracy(self):
        """Test accuracy matches expected values from Rust tests - mirrors check_itrend_accuracy"""
        expected = EXPECTED_OUTPUTS['ehlers_itrend']
        
        result = my_project.ehlers_itrend(
            self.close,
            expected['default_params']['warmup_bars'],
            expected['default_params']['max_dc_period']
        )
        
        assert len(result) == len(self.close)
        
        
        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=0,
            atol=0.1,  
            msg="Ehlers ITrend last 5 values mismatch"
        )
        
        
        compare_with_rust('ehlers_itrend', result, 'close', expected['default_params'])
    
    def test_ehlers_itrend_error_empty_input(self):
        """Test error with empty input - mirrors check_itrend_no_data"""
        with pytest.raises(ValueError, match="Input data is empty"):
            my_project.ehlers_itrend(np.array([]), 12, 50)
    
    def test_ehlers_itrend_error_all_nan(self):
        """Test error with all NaN values - mirrors check_itrend_all_nan_data"""
        all_nan = np.full(100, np.nan)
        with pytest.raises(ValueError, match="All values are NaN"):
            my_project.ehlers_itrend(all_nan, 12, 50)
    
    def test_ehlers_itrend_error_insufficient_data(self):
        """Test error with insufficient data for warmup - mirrors check_itrend_small_data_for_warmup"""
        small_data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        with pytest.raises(ValueError, match="Not enough data for warmup"):
            my_project.ehlers_itrend(small_data, 10, 50)
    
    def test_ehlers_itrend_error_zero_warmup(self):
        """Test error with zero warmup bars - mirrors check_itrend_zero_warmup"""
        with pytest.raises(ValueError, match="Invalid warmup_bars"):
            my_project.ehlers_itrend(self.close, 0, 50)
    
    def test_ehlers_itrend_error_invalid_max_dc(self):
        """Test error with invalid max_dc_period - mirrors check_itrend_invalid_max_dc"""
        with pytest.raises(ValueError, match="Invalid max_dc_period"):
            my_project.ehlers_itrend(self.close, 12, 0)
    
    def test_ehlers_itrend_reinput(self):
        """Test applying indicator twice - mirrors check_itrend_reinput"""
        expected = EXPECTED_OUTPUTS['ehlers_itrend']
        params = expected['default_params']
        
        
        first_result = my_project.ehlers_itrend(
            self.close,
            params['warmup_bars'],
            params['max_dc_period']
        )
        assert len(first_result) == len(self.close)
        
        
        second_result = my_project.ehlers_itrend(
            first_result,
            params['warmup_bars'],
            params['max_dc_period']
        )
        assert len(second_result) == len(first_result)
        
        
        
        
        if len(second_result) > 24:
            assert_no_nan(second_result[24:], "Found unexpected NaN after double warmup")
    
    def test_ehlers_itrend_nan_handling(self):
        """Test NaN handling - mirrors check_itrend_nan_handling"""
        expected = EXPECTED_OUTPUTS['ehlers_itrend']
        params = expected['default_params']
        warmup_bars = params['warmup_bars']
        
        result = my_project.ehlers_itrend(
            self.close,
            params['warmup_bars'],
            params['max_dc_period']
        )
        
        assert len(result) == len(self.close)
        
        
        assert_all_nan(
            result[:warmup_bars],
            msg="Expected NaN in warmup period"
        )
        
        
        if len(result) > warmup_bars:
            assert_no_nan(result[warmup_bars:], "Found unexpected NaN after warmup")
            
        
        assert warmup_bars == 12, f"Expected warmup_bars=12, got {warmup_bars}"
    
    def test_ehlers_itrend_streaming(self):
        """Test streaming interface - mirrors check_itrend_streaming"""
        
        batch_result = my_project.ehlers_itrend(self.close, 12, 50)
        
        
        stream = my_project.EhlersITrendStream(12, 50)
        stream_result = []
        
        for price in self.close:
            value = stream.update(price)
            stream_result.append(value if value is not None else np.nan)
        
        stream_result = np.array(stream_result)
        
        
        assert len(batch_result) == len(stream_result)
        
        
        
        for i in range(12, len(batch_result)):
            if not (np.isnan(batch_result[i]) and np.isnan(stream_result[i])):
                assert abs(batch_result[i] - stream_result[i]) < 1e-7, \
                    f"Batch/stream mismatch at index {i}: {batch_result[i]} vs {stream_result[i]}"
    
    def test_ehlers_itrend_batch_single_params(self):
        """Test batch calculation with single parameter set - mirrors check_batch_default_row"""
        expected = EXPECTED_OUTPUTS['ehlers_itrend']
        params = expected['default_params']
        
        
        batch_result = my_project.ehlers_itrend_batch(
            self.close[:100],  
            (params['warmup_bars'], params['warmup_bars'], 0),  
            (params['max_dc_period'], params['max_dc_period'], 0)  
        )
        
        
        assert isinstance(batch_result, dict)
        assert 'values' in batch_result
        assert 'warmups' in batch_result
        assert 'max_dcs' in batch_result
        
        
        assert batch_result['values'].shape == (1, 100)
        
        
        single_result = my_project.ehlers_itrend(self.close[:100], 12, 50)
        
        
        
        
        warmup_bars = params['warmup_bars']
        assert_close(
            batch_result['values'][0][warmup_bars:],
            single_result[warmup_bars:],
            rtol=0,
            atol=1e-10,
            msg="Batch vs single calculation mismatch after warmup"
        )
    
    def test_ehlers_itrend_batch_multiple_params(self):
        """Test batch calculation with multiple parameter combinations"""
        test_data = self.close[:100]
        batch_result = my_project.ehlers_itrend_batch(
            test_data,
            (10, 14, 2),      
            (40, 50, 10)      
        )
        
        
        assert batch_result['values'].shape[0] == 6
        assert batch_result['values'].shape[1] == 100
        
        
        assert len(batch_result['warmups']) == 6
        assert len(batch_result['max_dcs']) == 6
        
        
        expected_params = [
            (10, 40), (10, 50),
            (12, 40), (12, 50),
            (14, 40), (14, 50)
        ]
        
        for idx, (warmup, max_dc) in enumerate(expected_params):
            single_result = my_project.ehlers_itrend(test_data, warmup, max_dc)
            batch_row = batch_result['values'][idx]
            
            
            
            for i in range(warmup, len(single_result)):
                if not np.isnan(single_result[i]) and not np.isnan(batch_row[i]):
                    assert abs(batch_row[i] - single_result[i]) < 1e-9, \
                        f"Batch row {idx} (warmup={warmup}, max_dc={max_dc}) mismatch at index {i}"
            
            assert batch_result['warmups'][idx] == warmup
            assert batch_result['max_dcs'][idx] == max_dc
    
    def test_ehlers_itrend_batch_warmup_validation(self):
        """Test batch warmup period handling"""
        data = self.close[:30]
        
        batch_result = my_project.ehlers_itrend_batch(
            data,
            (10, 15, 5),      
            (50, 50, 0)       
        )
        
        
        assert batch_result['values'].shape == (2, 30)
        
        
        
        
        
        
        
        for i in range(10, 30):
            assert not np.isnan(batch_result['values'][0][i]) or \
                   np.isfinite(batch_result['values'][0][i]), \
                   f"Invalid value at index {i} in first row"
        
        
        for i in range(15, 30):
            assert not np.isnan(batch_result['values'][1][i]) or \
                   np.isfinite(batch_result['values'][1][i]), \
                   f"Invalid value at index {i} in second row"
    
    def test_ehlers_itrend_edge_cases(self):
        """Test edge cases for the indicator"""
        
        min_data = np.array([1.0] * 13)  
        result = my_project.ehlers_itrend(min_data, 12, 50)
        assert len(result) == 13
        
        
        assert_all_nan(result[:12], msg="Expected NaN in warmup period")
        
        
        test_cases = [
            (5, 30),   
            (20, 100), 
            (15, 15),  
        ]
        
        for warmup, max_dc in test_cases:
            if len(self.close) > warmup:
                result = my_project.ehlers_itrend(self.close[:100], warmup, max_dc)
                assert len(result) == 100, f"Failed for warmup={warmup}, max_dc={max_dc}"
                
                assert_all_nan(
                    result[:warmup],
                    msg=f"Expected NaN in warmup period for warmup={warmup}"
                )
    
    def test_ehlers_itrend_with_nan_input(self):
        """Test handling of data with NaN values in valid positions"""
        
        data_with_nan = self.close[:100].copy()
        data_with_nan[50:55] = np.nan
        
        
        result = my_project.ehlers_itrend(data_with_nan, 12, 50)
        assert len(result) == 100
        
        
        assert_all_nan(result[:12], msg="Expected NaN in warmup period")
        
        
        
        assert np.isnan(result[50]), "Expected NaN at index 50"
    
    def test_ehlers_itrend_performance_batch(self):
        """Test that batch processing is more efficient than multiple single calls"""
        import time
        
        test_data = self.close[:500]  
        
        
        start = time.time()
        batch_result = my_project.ehlers_itrend_batch(
            test_data,
            (10, 20, 2),  
            (40, 60, 10)  
        )
        batch_time = time.time() - start
        
        
        start = time.time()
        single_results = []
        for warmup in range(10, 21, 2):
            for max_dc in range(40, 61, 10):
                single_results.append(
                    my_project.ehlers_itrend(test_data, warmup, max_dc)
                )
        single_time = time.time() - start
        
        
        print(f"\n  Batch time: {batch_time:.3f}s, Single calls: {single_time:.3f}s")
        print(f"  Speedup: {single_time/batch_time:.1f}x")
        
        
        assert batch_result['values'].shape == (18, 500)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
